src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs

```diff
@@ -396,37 +418,28 @@ private CompressionMethodValues CompressionMethod
             }
         }
 
-        private string DecodeEntryName(byte[] entryNameBytes)
+        // Converts the specified bytes into a string of the proper encoding for this ZipArchiveEntry.
+        private string DecodeBytesToString(byte[] bytes)
         {
-            Debug.Assert(entryNameBytes != null);
-
-            Encoding readEntryNameEncoding;
-            if ((_generalPurposeBitFlag & BitFlagValues.UnicodeFileName) == 0)
+            Debug.Assert(bytes != null);
+
+            Encoding encoding;
+            // If the bit flag is not set, it either means we are adding a new entry, or
+            // this is an existing entry with ASCII comment and fullname.
+            // If the user chose a default encoding in the ZipArchive.ctor, then we use it, otherwise
+            // we default to UTF8.
+            if (!IsGeneralPurposeUnicodeFlagSet())
```

Comment:
```suggestion
            if (!_hasUnicodeEntryNameOrComment)
```

Reply:
Good find.

Reply:
This means we don't need to check `_generalPurposeBitFlag` anywhere from now on, we only check the `_hasUnicodeEntryNameOrComment` from now on, and the `_generalPurposeBitFlag` is only used when writing.

Reply:
I would prefer to use only on `_generalPurposeBitFlag` as IMO it is better to keep one source of truth rather than two.